src/utils/FixedQueue.ts

```diff
@@ -0,0 +1,52 @@
+import * as _ from 'underscore';
+import { deepClone } from '../../CloneUtils';
+
+export class FixedQueue<T> {
+
+  private queue: Array<T>;
+  private maxLength: number;
+
+  constructor(initialValues: Array<T> = [], maxLength: number = Number.MAX_VALUE) {
+    this.maxLength = maxLength;
+    this.queue = initialValues;
+    this.trimQueue();
+  }
+
+  immutablePush(object: any): FixedQueue<T> {
+    const queue: Array<T> = deepClone(this.queue);
+    queue.push(object);
```

Comment:
By sending `queue` to your new `FixedQueue`, you are still using the same reference. Honestly, this is getting a lot more complex than it needs to be. You are going through a maze of functions just to push an item into an array.